src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs

```diff
@@ -402,7 +406,7 @@ private static IEnumerable<MsalAccessTokenCacheItem> FilterByTokenType(Authentic
             string requestTenantId = requestParams.Authority.TenantId;
             bool filterByTenantId = true;
 
-            if (requestParams.UserAssertion != null) // OBO
+            if (requestParams.ApiId == ApiEvent.ApiIds.AcquireTokenOnBehalfOf) // OBO
             {
```

Comment:
nit: are we sure this is getting set? & if we check on the UserAssertion, we are sure it's not null

Reply:
because if you change the condition, and the assertion is null, you'll get a null ref later.

Reply:
Same condition is used below [see](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/92ed27f24009f1b1574d4b5fdc5522be4819bfc8/src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs#L440). `UserAssertion` here is used to check for OBO flow. Discussed it with @bgavrilMS. I can revert it if required.

Reply:
We are sure that this param (ApiId) is always is set and we make this kind of verification is many other places. 

I think the code is more readable like this. In fact, UserAssertion type is used in IWA and ROPC as well, but it is not set in the request params, so this code path is followed exclusively in OBO. It took me 5 min to figure this out :( since we are creating a UserAssertion in ~7 places.

Reply:
We are sure that this param (ApiId) is always is set and we make this kind of verification is many other places.

I think the code is a bit more readable like this. UserAssertion is also used in IWA and ROPC , but not for cache operations!! This code path is followed exclusively in OBO. It took me 5 min to figure this out :( since we are creating a UserAssertion object in ~7 places, but only OBO sets it in the requestParams. 

So either we say - if this request has a UserAssertion, then filter by it or we say "in case of OBO, filter by user assertion". IWA and ROPC use UserAssertiom, but do not fitler by it. Hence it makes a bit more sense to have the code as Neha changed it.

Reply:
that wasn't my point, sorry i wasn't clear. If we remove the check for UserAssertion being null, and then use the UserAssertion on line 413, that will throw a null ref if it's null: `requestParams.UserAssertion.AssertionHash`. so we need to check if it's null before we use it. that's all.

Reply:
We don't allow OBO without a UserAssertion ... I don't think. We need to bail otherwise, because OBO does not have information about the user, i.e. our filtering would be completely screwed up if we had OBO and null UserAssertion.

Maybe log an error?

Reply:
or we just check that it's done earlier in the code, i haven't checked and add a comment so we know for future reference that it's not null. just seems odd to be checking this late, i agree.

Reply:
Took a look at the code path. We can add a null check here at the earliest. The parameter is required, but null can still be passed in, so we could get NPE in token cache. https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/f2866fc6884a8a4bb3a210dc1150b3f6088d3e96/src/client/Microsoft.Identity.Client/ConfidentialClientApplication.cs#L122-L129

Reply:
I don't think this will scenario will ever occur where there will not be a user assertion for an obo flow as it is a required param for AcquireTokenOnBehalfOf.

Reply:
Added check